src/raft.c

```diff
@@ -178,8 +137,57 @@ static ShardGroup *GetSlotShardGroup(RedisRaftCtx *rr, unsigned int slot, bool a
     return sg;
 }
 
-static RRStatus validateRaftRedisCommandArray(RedisRaftCtx *rr, RedisModuleCtx *reply_ctx,
-                                              RaftRedisCommandArray *cmds, unsigned int slot)
+static RRStatus validateKeyExistence(RedisModuleCtx *ctx,
+                                     RaftRedisCommandArray *cmds,
+                                     unsigned int slot,
+                                     SlotRangeType slot_type)
+{
+    int total_keys = 0, num_keys_found = 0;
+    RedisRaftCtx *rr = &redis_raft;
+
+    if (slot_type == SLOTRANGE_TYPE_STABLE) {
+        return RR_OK;
+    }
+
+    for (int i = 0; i < cmds->len; i++) {
+        int num_keys, *keyindex;
+        RaftRedisCommand *cmd = cmds->commands[i];
+
+        /* Iterate command keys */
+        keyindex = RedisModule_GetCommandKeys(rr->ctx, cmd->argv, cmd->argc, &num_keys);
+        total_keys += num_keys;
+
+        for (int j = 0; j < num_keys; j++) {
+            int nokey; /* Key should not be locked */
+            RedisModule_DictGet(rr->locked_keys, cmd->argv[keyindex[j]], &nokey);
+
+            if (nokey && RedisModule_KeyExists(rr->ctx, cmd->argv[keyindex[j]])) {
+                num_keys_found++;
+            }
+        }
+        RedisModule_Free(keyindex);
+    }
+
+    if (num_keys_found == total_keys) {
+        return RR_OK;
+    }
+
+    /* There are missing keys, reply an error */
+    if (ctx) {
+        if (num_keys_found == 0 && slot_type == SLOTRANGE_TYPE_MIGRATING) {
+            replyAsk(rr, ctx, slot);
+        } else {
+            RedisModule_ReplyWithError(ctx, "TRYAGAIN");
+        }
+    }
```

Comment:
this isn't correct.  num_keys_found can == 0 with a locked key.  we don't reply ASK then.

Reply:
oops, right